20250429-RS-scheduler-broadcast-job-type.md

```diff
@@ -0,0 +1,320 @@
+Moving this functionality as a first concept to Dapr improves performance and improves the developer experience in creating complex distributed systems.
+
+### Design
+
+The Scheduler service will implement a new `Broadcast` typed job.
+This new Job type is an extension on the existing "typical" `Job` type, with the difference being that instead of the Job being invoked at some schedule to a _single_ client, the Job itself will be Broadcast to _all_ clients which implement that type.
+
+The Scheduler Job handler will be updated to special case these Broadcast job types to be extended with the following functionality:
+1. Instead of Scheduler sending the Job trigger event to a single client, it will be sent to all clients which implement that job type.
+2. Upon connection of a client which implements that type, Scheduler will send the job to each implementing client with the `BROADCAST_JOB_EVENT_PUT` flag set.
+3. For each Job trigger event, each client will receive the event with the `BROADCAST_JOB_EVENT_TRIGGER` flag set.
+4. At any point during the Job's life cycle it is deleted (not expired according to its schedule), the Scheduler will send a `BROADCAST_JOB_EVENT_DELETE` event to all clients which implement that job type.
+
+## Implementation Details
+
+#### go-etcd-cron
+
+In order for Scheduler to track the existing Broadcast jobs which are created, exist during a restart, or are deleted, a new hook will be implemented in the cron Informer.
+This hook will be called during informer events, and the Scheduler will use these events to manage the in memory Broadcast job state and manage broadcasting those events.
+
+#### Scheduler
+
+The Scheduler service will be updated to include a new Job handler code path to build a Broadcast Job manager.
+This manager will take informer events from go-etcd-cron, sending PUT and DELETE flagged job events to connected clients during CRUD events, as well as when appropriate clients connect.
+The Scheduler will trigger all appropriate connected clients concurrently during trigger events.
+It is not expected that clients would return any response code to a trigger execution other than SUCCESS, as these jobs are generally intended (at least for the initial use cases) to provide object manifests- not triggering some execution.
+A client failing to return the response of a trigger should be considered as a failed connection.
+In any case, to account for future use cases, the trigger response codes should considered a logical AND whereby failed calls will be retired on each failed client call according to the Jobs Failure Policy.
+
+Broadcast jobs can either expire according to its schedule, or be deleted by a client.
+
+Broadcast Jobs will continue to adhere to the staging queue, like existing Jobs today.
+
+#### Keep Alive
+
+The gRPC keep alive options for both the Scheduler server and daprd clients need to be updated so that they are more aggressive.
```

Comment:
why does this need to change?

Reply:
To ensure we have better settings to ensure the health of each stream to Scheduler. This was written more as a TODO for myself to improve the integrity of the client connection pool state. Realistically, this shouldn't have any behavioral changes to the API.